Event system RFC

ericvicenti opened the pull request:
Resume discussion of event system for screen/navigator focus in React Navigation!

brentvatne commented on 0000-event-system.md:
````diff
@@ -0,0 +1,109 @@
+```
+componentDidMount() {
+	this._sub = this.props.navigation.events.addListener(
+		'DidFocus',
````
- I think event names should be camelcase with the first letter starting as lower case -- `didFocus` -- this is more of a standard for events in JS (eg: https://reactjs.org/docs/events.html#reference)
- Additionally, I think I personally prefer for the events to use the `on` prefix, such as `onFocus`, `onBlur`, `onWillFocus` (once again for consistency, see above link again)
- Lastly, I think we should hoist the `events.addListener` function up to the `navigation` object itself -- so we have `this.props.navigation.addEventListener`(`addEventListener` is used by AppState and NetInfo so the name is consistent). This is a smaller API surface area --
 only one property is exposed, `addEventListener`, rather than both `events` and `addListener` on events.

ericvicenti replied:
Sure, these conventions seem more common. I'll fix up the RFC for that

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+Screens need to know about navigation changes while they are mounted. Consider the following use cases:
+
+### Data and Media subscription
+Screens must know when to request live data or resource-intensive APIs. Media should be able to auto-pause at correct times.
```
**and when to enable/disable** resource-intensive APIs

ericvicenti replied:
sure, will add this

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+Screens must know when to request live data or resource-intensive APIs. Media should be able to auto-pause at correct times.
+
+### Unmounting
+Tabs and cards may need to unmount and re-mount contents when they are focused.
```
worth mentioning memory usage concern here, and integration with `removeClippedSubviews` via https://github.com/SoftwareMansion/react-native-resource-saving-container

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+Tabs and cards may need to unmount and re-mount contents when they are focused.
+
+### Back button handling
+Screens may want to prevent back actions based on internal state
```
I think it's rare for this to actually happen on iOS, it's more of an Android concern because you can press a back button at any time. On iOS if the back button is visible and not greyed out, and you can swipe back, then you should be able to go back. We should call out Android specifically here

ericvicenti replied:
Ok I can definitely call out Android here. But I would note that this Android feature has a similarity on iOS: the back gesture being enabled or not. Right now we have an independent mechanism to control that, but it is super painful to use when it depends on internal screen state.

brentvatne replied:
oh interesting! thanks for pointing that out. we should mention it here too

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+Screens may want to prevent back actions based on internal state
+
+### Keyboard dismissal
+KB must dismiss when scene *starts* to transition back (mayBlur)
```
Another one that is important here is StatusBar, where if it needs to be light on one screen and dark on another, we want to use a sane default for when to change the status bar automatically but let people control it more with route events if they need to

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+### Keyboard dismissal
+KB must dismiss when scene *starts* to transition back (mayBlur)
+
+## Input autofocus
```
looks like a `#` is missing in this header

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+
+# Unresolved questions
+
+- How should the implementation work?
```
could you elaborate on this question?

ericvicenti replied:
Specifically I'm wondering what module should be responsible for providing the events

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,109 @@
+- How should the implementation work?
+- How should back button cancellation work?
+- Refocus event vs extra dispatches of 'Focus'
+- Should there be a general 'route change' event, or will that simply be handled via props.navigation.state?
```
can you elaborate on this question as well?

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,142 @@
+
+In some cases, the navigation state has not changed and `props.navigation.state === newState`. One example of this is the 'onMaySwipeBack' action fired by the stack navigator.
+
+The following events will be emitted _inside of navigators_. Navigators will implement these additional actions by subscribing to its own onAction event from `props.navigation.addEventListener`.
```
might want to clarify what _inside of navigators_ means

brentvatne commented on 0000-event-system.md:
```diff
@@ -0,0 +1,142 @@
+
+### onAction(action, newState)
+
+This fires for every action that may affect the navigation state.
```
if i'm listening for an action in the profile route within a stack in the home tab, and some navigation action is dispatched from the settings tab, would this fire?

ericvicenti replied:
nope and I clarified this

satya164 commented on 0000-event-system.md:
````diff
@@ -0,0 +1,142 @@
+```
+componentDidMount() {
+  this._sub = this.props.navigation.addEventListener(
+    'onDidFocus',
````
I think it'll be better name these events as just `didFocus`, `didBlur` etc. without the `on` prefix:

`onDidFocus` -> a prop to a react component, or dom element etc., similar to `onClick`, or a property on a element instance e.g. - `socket.onopen = () => { ... }`
`didFocus` -> similar to `addEventListener('click', () => { ... })` etc.

brentvatne replied:
@satya164 - naming was inspired by https://reactjs.org/docs/events.html#reference

satya164 replied:
@brentvatne those names are name of the props to components, e.g. `<button onClick>` and similar to `<button onclick>` in HTML, I think these names don't fit well with an `addEventListener` API which drops the `on` prefix

satya164 commented on 0000-event-system.md:
````diff
@@ -0,0 +1,140 @@
+
+```
+componentDidMount() {
+  this._sub = this.props.navigation.addEventListener(
````
Maybe `addListener` fits better because of several React Native APIs such as `Keyboard`, `Animated` etc. using `addListener`. It also signals that the behaviour is different from web's `addEventListener` which doesn't return a subscription object.

brentvatne replied:
alas `addEventListener` is used by `NetInfo` and `AppState`, there is no consistency in the world, sigh

ericvicenti commented:
ok I did as we discussed:
- removed the "on" from "onDidFocus"
- removed the "Event" from "addEventListener"
- added "removeListener"

Ready to land the proposal!

sibelius commented:
missing something like this https://github.com/react-navigation/react-navigation/issues/1363#issuecomment-356340978

brentvatne commented:
@sibelius - not sure what you mean. can you explain in more detail?
